physicsnemo/sym/eq/ls/grads.py

```diff
@@ -23,148 +23,189 @@
+
+            all_dv = torch.zeros(
+                num_nodes, max_neighbors, self.dim, device=coords.device
             )
+            all_du = torch.zeros(num_nodes, max_neighbors, 1, device=coords.device)
+            all_weights = torch.zeros(num_nodes, max_neighbors, device=coords.device)
+
+            for i in range(num_nodes):
+                start_idx = offsets[i]
+                end_idx = offsets[i + 1]
+                neighbor_indices = indices[start_idx:end_idx]
+
+                if len(neighbor_indices) == 0:
+                    continue
+
+                p_center = coords[i : i + 1]  # [1, dim]
+                p_neighbors = coords[neighbor_indices]  # [num_neighbors, dim]
+
+                f_center = y[i : i + 1]  # [1, 1]
+                f_neighbors = y[neighbor_indices]  # [num_neighbors, 1]
+
+                dv = p_neighbors - p_center  # [num_neighbors, dim]
+                du = f_neighbors - f_center  # [num_neighbors, 1]
+
+                weights = 1 / (torch.sum(dv**2, dim=1) + 1e-8)  # [num_neighbors]
+
+                num_neighbors = len(neighbor_indices)
+                all_dv[i, :num_neighbors] = dv
+                all_du[i, :num_neighbors] = du
+                all_weights[i, :num_neighbors] = weights
+
+            grad_u = self.compute_ls_grads(all_dv, all_du)  # [N, dim, 1]
+            grad_u = grad_u.squeeze(-1)  # [N, dim]
+
+            # Split into individual components
+            result = []
+            for i in range(self.dim):
+                result.append(grad_u[:, i : i + 1])  # [N, 1]
 
-            result.append(dudx.unsqueeze(dim=1))
-            result.append(dudy.unsqueeze(dim=1))
-            result.append(dudz.unsqueeze(dim=1))
             return result
+
+    def _forward_batched(self, coords, neighbor_matrix, y) -> List[Tensor]:
+        """
+        Compute derivatives using batched connectivity format.
+        """
+        num_nodes = coords.shape[0]
+        max_neighbors = neighbor_matrix.shape[1]
+
+        # Create mask for valid neighbors
+        valid_mask = neighbor_matrix != -1  # [N, max_neighbors]
+
+        neighbor_coords = coords[neighbor_matrix]  # [N, max_neighbors, dim]
+        neighbor_values = y[neighbor_matrix]  # [N, max_neighbors, 1]
+
+        center_coords = coords.unsqueeze(1)  # [N, 1, dim]
+        center_values = y.unsqueeze(1)  # [N, 1, 1]
+
+        dv = neighbor_coords - center_coords  # [N, max_neighbors, dim]
+        du = neighbor_values - center_values  # [N, max_neighbors, 1]
+
+        # Apply mask to zero out invalid neighbors
+        mask_expanded = valid_mask.unsqueeze(-1)  # [N, max_neighbors, 1]
+        dv = dv * mask_expanded
+        du = du * mask_expanded
+
+        grad_u = self.compute_ls_grads(dv, du)  # [N, dim, 1]
+        grad_u = grad_u.squeeze(-1)  # [N, dim]
+
+        # Split into individual components
+        result = []
+        for i in range(self.dim):
```

Comment:
Same as in another place:

Not a critical change, but for best-practices, this should be rewritten with a list comprehension. (For example, Ruff would flag this as a performance error)